File: 2026-02-February-GFG-POTD/February 02 - Max Circular Subarray Sum/Solution.cpp
class Solution {
public:
    int maxCircularSum(int arr[], int n) {
        int totalSum = arr[0];

        int maxEndingHere = arr[0];
        int maxSoFar = arr[0];

        int minEndingHere = arr[0];
        int minSoFar = arr[0];

        for (int i = 1; i < n; i++) {
            int x = arr[i];

            // Kadane for max subarray sum
            maxEndingHere = max(x, maxEndingHere + x);
            maxSoFar = max(maxSoFar, maxEndingHere);

            // Kadane for min subarray sum
            minEndingHere = min(x, minEndingHere + x);
            minSoFar = min(minSoFar, minEndingHere);

            totalSum += x;
        }

        // If all elements are negative
        if (maxSoFar < 0)
            return maxSoFar;

        // Maximum of wrapping and non-wrapping
        return max(maxSoFar, totalSum - minSoFar);
    }
};

File: 2026-02-February-GFG-POTD/February 02 - Max Circular Subarray Sum/Solution.py
class Solution:
    def maxCircularSum(self, arr):
        totalSum = arr[0]

        maxEndingHere = arr[0]
        maxSoFar = arr[0]

        minEndingHere = arr[0]
        minSoFar = arr[0]

        for i in range(1, len(arr)):
            x = arr[i]

            # Kadane for max subarray sum
            maxEndingHere = max(x, maxEndingHere + x)
            maxSoFar = max(maxSoFar, maxEndingHere)

            # Kadane for min subarray sum
            minEndingHere = min(x, minEndingHere + x)
            minSoFar = min(minSoFar, minEndingHere)

            totalSum += x

        # If all elements are negative
        if maxSoFar < 0:
            return maxSoFar

        # Maximum of wrapping and non-wrapping
        return max(maxSoFar, totalSum - minSoFar)
